### PR TITLE
Add option for generating RBIs for Rails engines

### PR DESCRIPTION
**Pull request by andyw8**
(Paired with @vinistock) 

### Motivation

Closes https://github.com/Shopify/tapioca/issues/1182

### Implementation

An `--app_root` option was added to the `tapioca dsl` command. This can be used to specify the path to the dummy app for an engine.

There were several places in the code which assumed the app is always at the root of the project. These were updated to take the `app_root` option into consideration.

### Tests

The test approach is styled after the existing RBI generation tests. We create files for the app as normal, but then move the necessary files into the `test/dummy` path before running the `dsl` commmand.

### CODE REVIEW

File: spec/tapioca/cli/dsl_spec.rb [hunk 1]
```diff
@@ -343,6 +343,62 @@ def title=(title); end
...
+                def title; end
+
+                sig { params(title: T.nilable(::String)).returns(T.nilable(::String)) }
+                def title=(title); end
+              end
+            end
+          RBI
+
+          # Restore directory structure so to not impact other tests
+          FileUtils.mv(engine_path + "/config", @project.path)
```
Comment (vinistock) [inline spec/tapioca/cli/dsl_spec.rb:397]: Can we put this inside an `ensure` block? It makes sure that we restore back even if something fails.

Comment (andyw8) [reply spec/tapioca/cli/dsl_spec.rb]: Good call, I've pushed an update.

File: spec/tapioca/cli/dsl_spec.rb [hunk 2]
```diff
@@ -343,6 +343,66 @@ def title=(title); end
...
+              end
+            end
+          RB
+
+          engine_path = @project.path + "/test/dummy"
+
+          begin
+            FileUtils.mkdir_p(engine_path)
+            FileUtils.mv(@project.path + "/config", engine_path)
+            FileUtils.mv(@project.path + "/lib", engine_path)
```
Comment (KaanOzkan) [inline spec/tapioca/cli/dsl_spec.rb:372]: I'd prefer writing the files to `/test/dummy` directly on line 347 but not a blocker.

**Review by KaanOzkan — APPROVED** [review]

File: spec/tapioca/cli/dsl_spec.rb (hunk 2 above)
Comment (andyw8) [reply spec/tapioca/cli/dsl_spec.rb:372]: Agreed, unfortunately the top-level `before(:all)` sets the project path so we would need to restructure the tests first.

Comment (KaanOzkan) [reply spec/tapioca/cli/dsl_spec.rb:372]: I'm a bit confused. I was only referring to the `lib` files. More specifically I was thinking of changing the 2 writes to:
`@project.write("test/dummy/lib/post.rb", <<~RB)` & `@project.write("test/dummy/lib/comment.rb", <<~RB)`

and removing the line `FileUtils.mv(@project.path + "/lib", engine_path)`.

We don't have to do it, I'm just trying to understand the project path blocker 🙂

**Pull request merged**